budget-api/src/services/transactions.service.ts

```diff
@@ -34,30 +37,42 @@ export class TransactionsService {
   }
 
   static async addMany(transactions: TransactionDto[]): Promise<Transaction[]> {
-    const allTagIs = transactions.reduce((ts, tran) => {
-      return tran?.tagIds?.length > 0 ? [...ts, ...tran.tagIds] : [];
+    const tagNames = transactions.reduce((tn, tran) => {
+      return tran?.tagNames?.length > 0 ? [...tn, ...tran.tagNames] : [...tn];
     }, []);
+    const transactionTags = await TagsService.addMany(tagNames);
+    const tagsMap = new Map(transactionTags.map((tag) => [tag.name, tag]));
 
-    const allTags = await TagsService.getManyById([...new Set(allTagIs)]);
+    const sourceNames = [...new Set(transactions.map((tran) => tran?.sourceName))];
+    const transactionSources = await SourcesService.getManyByNames(sourceNames);
+    const sourcesMap = new Map(transactionSources.map((source) => [source.name, source]));
 
-    const transactionArray = transactions.map((tran) =>
-      TransactionConverter.fromDto(
-        tran,
-        allTags.filter((tag) => tran.tagIds.includes(tag.id))
-      )
-    );
+    if (sourceNames.length > transactionSources.length) {
+      const missing = sourceNames.filter(name => !sourcesMap.has(name));
+      throw new NotFoundError(`The following sources were not found: '${missing.join("', '")}'`);
+    }
+
+    const transactionArray = transactions.map((tran) => {
+      const source = sourcesMap.get(tran.sourceName) as Source;
+      const tags = tran.tagNames.map((name) => tagsMap.get(name));
+      return TransactionConverter.fromDto(tran, source, tags)
+    });
 
     return this.saveMultipleTransactions(transactionArray);
   }
 
-  static async updateOne(data: TransactionDto): Promise<Transaction> {
-    const updatedTransaction = await Transaction.findOne({
-      id: data.transactionId,
-    });
+  static async updateOne(id: number, data: TransactionDto): Promise<Transaction> {
+    const updatedTransaction = await Transaction.findOne({ id });
```

Comment:
It's somewhat dangerous to use `{ id }` as a parameter for `findOne`: if the `id` parameter of the `updateOne` method is renamed (say, using F2), this search for one transaction will stop working.
My suggestion is to rename the `id` parameter of this method right away and also to pass `{ id: <new_parameter_name >}` object into `findOne` here.